tiledb/sm/storage_manager/context.h

```diff
@@ -180,12 +181,27 @@ class Context {
   size_t get_io_thread_count(const Config& config);
 
   /**
-   * Initializes global and local logger.
+   * Get the logging format to use
    *
    * @param config The configuration parameters.
-   * @return Status
+   * @return The log format to use
    */
-  Status init_loggers(const Config& config);
+  Logger::Format get_log_format(const Config& config) noexcept;
+
+  /**
+   * Get the logging level to use
+   *
+   * @param config The configuration parameters.
+   * @return The log level to use
+   */
+  Logger::Level get_log_level(const Config& config) noexcept;
+
+  /**
+   * Log any issues with the logging configuration
+   *
+   * @param config The Config instance to check
+   */
+  void check_logger_config(const Config& config) noexcept;
```

Comment:
So the question I was dealing with here is how do you communicate an error while configuring the thing we use to communicate those errors. If a log config is broken I have to assume its before we've even configured the global logger so that's out.

You're quite right that the existing approach abuses the lack of C.41 consistency in Logger to just log the errors it encounters while configuring the logger which is more or less what I went for. In the face of any error, default to constructing an ERR level, DEFAULT formatted log so that the errors can be printed.

My main worry here is that being *more* C.41 compliant by not using the thing until its properly constructed, means we can't communicate any errors outside of an opaque `TILEDB_ERR` which means if we throw before any logger has been configured, users won't have any idea why whatever operation failed just because they typed "json" in lower case in their config file.

So I did change things a bit here in that its now "Absolutely get a logger configured at least into the ERR/DEFAULT level/format state so that we can spit out actual error messages for users. The assumption being that if someone really is concerned about their logging config and looking at the file, they'll have a message as to why its not what was expected.

We could also go another route in that the global Logger singleton is instantiated purely off OS environment variables such that the configuration can be checked at libtiledb's library initialization time, but even then I would very much expect that to be a safe fallback situation just the same as we have where if there's a config issue, we still create a logger and just log the issue rather than do... something else?

Which is to say, normally, I'd agree that if we're configuring/creating something and detect configuration issues the right thing is to stop and report the issue rather than blindly follow on. However, if the thing we're configuring is the thing we use to communicate the issue then I'd consider it more important to get the communication channel open and report after the fact.

Reply:
I went to implement the LoggerParameters struct similar to the VFSParameters struct and realized that it'll do the same thing to this PR as my first approach. Having Logger require a link dependency against Config means that we're rewriting a good chunk of the build system to move things around so that we're linking everything properly.

I don't think that's necessarily a bad idea in general, I just think its beyond the scope of this PR which is supposed to just be "make these trace log statements actually emit log messages in one very specific bit of code".

Reply:
In fact, when I put it that way, we could actually make this entire thing just that, pass a log level to the constructor for now without changing anything else at all.

https://github.com/TileDB-Inc/TileDB/compare/dev...pd/sc-25861/fix-logger-init-order-2